app/scripts/metamask-controller.js

```diff
@@ -5118,6 +5123,58 @@ export default class MetamaskController extends EventEmitter {
     );
   }
 
+  setUpCookieHandlerCommunication({ connectionStream }) {
+    const {
+      metaMetricsId,
+      dataCollectionForMarketing,
+      participateInMetaMetrics,
+    } = this.metaMetricsController.store.getState();
+
+    if (
+      metaMetricsId &&
+      dataCollectionForMarketing &&
+      participateInMetaMetrics
+    ) {
+      // setup multiplexing
+      const mux = setupMultiplex(connectionStream);
+      const metamaskCookieHandlerStream = mux.createStream(
+        METAMASK_COOKIE_HANDLER,
+      );
+      // set up postStream transport
+      // TODO: Add handler here to modify trackEvents metrics context
+      // Ideally take ga_client_id here from page message
+      metamaskCookieHandlerStream.on(
+        'data',
+        createMetaRPCHandler(
+          {
+            getCookieFromMarketingPage:
+              this.getCookieFromMarketingPage.bind(this),
+          },
+          metamaskCookieHandlerStream,
+        ),
+      );
+    }
+  }
+
+  getCookieFromMarketingPage(data) {
+    const { ga_client_id: cookieId } = data;
+    const {
+      metaMetricsId,
+      dataCollectionForMarketing,
+      participateInMetaMetrics,
+    } = this.metaMetricsController.store.getState();
+
+    if (
+      metaMetricsId &&
+      dataCollectionForMarketing &&
+      participateInMetaMetrics
```

Comment:
Question: `getCookieFromMarketingPage` is not called unless there are defined and true in `setUpCookieHandlerCommunication`, so in which condition this will be called ?
Hence we could remove this and directly call `this.metaMetricsController.setMarketingCampaignCookieId`

Reply:
`setUpCookieHandlerCommunication` will be called only when we set up the stream communication. Each time the message is triggered, `getCookieFromMarketingPage` will be called. Even when we turn off `participateInMetrics` or `dataCollectionForMarketing`, `getCookieFromMarketingPage` will be called, so we need a check inside the method. 

But this has me thinking if we should remove the check inside `setUpCookieHandlerCommunication` since the stream might not be set up if `participateInMetrics` or `dataCollectionForMarketing` is not true initially but turned on later, I will test out this scenario.

Reply:
If you check this line https://github.com/MetaMask/metamask-extension/blob/32f81559e47fabc7bcb02ae18a7310074cd8ac05/app/scripts/metamask-controller.js#L5133 and actually `getCookieFromMarketingPage` won't be called until all conditions are satisfied.

Reply:
That line is to set up the stream listener, which happens during page loading; if a message event occurs later, `getCookieFromMarketingPage` is triggered, so I just added that to be on the safe side, in case metrics or data collection status changes in between.

Reply:
After testing with the script provided by marketing team, I did updated this.